nrf52-hal-common/src/uarte.rs

```diff
@@ -359,9 +340,530 @@ impl<T> Uarte<T> where T: Instance {
     pub fn free(self) -> T {
         self.0
     }
+
+    /// Splits the UARTE into an interrupt driven transmitter and receiver
+    ///
+    /// In needs a `Consumer` to place DMA chunks to send, a `Timer` for checking byte timeouts,
+    /// and a `&'static mut [u8]` memory block to use.
+    pub fn into_interrupt_driven<S, I>(
+        self,
+        txc: Consumer<'static, Box<UarteDMAPool>, S>,
+        timer: Timer<I>,
+        dma_pool_memory: &'static mut [u8],
+    ) -> (UarteRX<T, I>, UarteTX<T, S>)
+    where
+        S: ArrayLength<heapless::pool::singleton::Box<UarteDMAPool>>,
+        I: timer::Instance,
+    {
+        debug_assert!(
+            dma_pool_memory.len() >= core::mem::size_of::<UarteDMAPoolNode>() * 2,
+            "The memory pool needs at least space for 2 `UarteDMAPoolNode`s"
+        );
+        UarteDMAPool::grow(dma_pool_memory);
+
+        let rx = UarteRX::<T, I>::new(timer);
+        let tx = UarteTX::<T, S>::new(txc);
+        (rx, tx)
+    }
 }
 
-impl<T> fmt::Write for Uarte<T> where T: Instance {
+/// An interrupt driven implementation of the UARTE peripheral.
+///
+/// ## Example
+///
+/// ```
+///
+/// ```
+pub mod interrupt_driven {
+    use core::sync::atomic::{compiler_fence, Ordering::SeqCst};
+    use core::{fmt, mem, ptr, slice};
+
+    use crate::timer::{self, Timer};
+    use crate::uarte;
+    use embedded_hal::timer::{Cancel, CountDown};
+
+    use heapless::{
+        consts, pool,
+        pool::singleton::{Box, Pool},
+        spsc::{Consumer, Queue},
+        ArrayLength,
+    };
+
+    /// DMA block size
+    /// Defaults to UARTE_DMA_SIZE = 16 if not explicitly set
+    #[cfg(not(any(
+        feature = "UARTE_DMA_SIZE_4",
+        feature = "UARTE_DMA_SIZE_8",
+        feature = "UARTE_DMA_SIZE_16",
+        feature = "UARTE_DMA_SIZE_32",
+        feature = "UARTE_DMA_SIZE_64",
+        feature = "UARTE_DMA_SIZE_128",
+        feature = "UARTE_DMA_SIZE_255"
+    )))]
+    pub const UARTE_DMA_SIZE: usize = 16;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_4")]
+    pub const UARTE_DMA_SIZE: usize = 4;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_8")]
+    pub const UARTE_DMA_SIZE: usize = 8;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_16")]
+    pub const UARTE_DMA_SIZE: usize = 16;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_32")]
+    pub const UARTE_DMA_SIZE: usize = 32;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_64")]
+    pub const UARTE_DMA_SIZE: usize = 64;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_128")]
+    pub const UARTE_DMA_SIZE: usize = 128;
+
+    // Currently causes internal OOM, needs fixing
+    // Maximum DMA size is 255 of the UARTE peripheral, see `MAXCNT` for details
+    #[cfg(feature = "UARTE_DMA_SIZE_255")]
+    pub const UARTE_DMA_SIZE: usize = 255;
```

Comment:
I don't like this. What if multiple in the dependency graph depend on the HAL, but specify different sizes? Then it won't compile.

This should be added as a generic argument on `UarteDMAPoolNode` and wherever else it's needed.

Reply:
Agreed its ugly. If I recall it needs to be a constant as we are using the size for allocation. Not sure if we can do it currently in Rust in a more elegant way. I'm open to suggestions here.

Reply:
Ah right. I remember there are limitations around array lengths (although I don't recall the details), so it makes sense you needed to do this to get it to work.

I'm fine with merging it like this for now. Const generics seem to be making progress lately, so hopefully we can fix it by the time someone runs into the dependency graph problems I described.

Reply:
Its used to allocate memory, not sure if we have const eval for usize.

Reply:
Well, if you use typenum the value is evaluated at compile time so it should be zero cost.

Reply:
If `typenum` is what `heapless` uses, then we should do that.

Reply:
`heapless` uses `generic-array` which uses`typenum`.